src/NHibernate/Util/WeakHashtable.cs

```diff
@@ -131,124 +117,154 @@ public bool MoveNext()
 
 		public void Reset()
 		{
-			innerEnumerator.Reset();
-			currentKey = currentValue = null;
+			_innerEnumerator.Reset();
+			_current = default(KeyValuePair<object, object>);
+		}
+
+		public KeyValuePair<object, object> Current => _current;
+
+		object IEnumerator.Current => _current;
+
+		#region IDisposable Support
+		private bool _disposedValue = false; // To detect redundant calls
+
+		protected virtual void Dispose(bool disposing)
+		{
+			if (!_disposedValue)
+			{
+				if (disposing)
+				{
+					_innerEnumerator.Dispose();
+					_current = default(KeyValuePair<object, object>);
+				}
+
+				_disposedValue = true;
+			}
 		}
 
-		public object Current
+		// This code added to correctly implement the disposable pattern.
+		public void Dispose()
 		{
-			get { return Entry; }
+			// Do not change this code. Put cleanup code in Dispose(bool disposing) above.
+			Dispose(true);
 		}
+		#endregion
 	}
 
 	[Serializable]
-	public class WeakHashtable : IDictionary
+	public class WeakHashtable : IDictionary<object, object>
```

Comment:
Minimal switch, the next commit turns this to a fully generic type `WeakHashtable<TKey, TValue>`.

But I question whether it is worth it or not. On Java side, they have ceased using this for query plan caching. (Java side, they have start using that weak hashtable for first level cache it seems.) They are using a 2048 max entries (configurable) plan cache instead with an algorithm for evicting coldest entries. And moreover, they use something more efficient than weak references for their remaining "caching" purpose. (Usage of weak reference for caching is by the way advises against by Microsoft documentation.) Weak references may be reclaimed while being far from reaching available memory limits. Java use soft references instead, designed to be reclaimed only when under memory stress. They do not have equivalent in .Net. 

We may have better throwing that away for using a `System.Runtime.Caching.MemoryCache` instead (available since fx4).